src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs

```diff
@@ -414,61 +415,119 @@ public bool TryCreateType(ITypeSymbol symbol, [NotNullWhen(true)] out CSharpType
         public bool TryCreateType(ITypeSymbol symbol, Func<System.Type, bool> validator, [NotNullWhen(true)] out CSharpType? type)
         {
             type = null;
-            INamedTypeSymbol? namedTypeSymbol = symbol as INamedTypeSymbol;
-            if (namedTypeSymbol == null)
+            return symbol switch
             {
-                throw new InvalidCastException($"Unexpected type {symbol}");
-            }
+                IArrayTypeSymbol arrayTypeSymbol => TryCreateTypeForIArrayTypeSymbol(arrayTypeSymbol, validator, out type),
+                INamedTypeSymbol namedTypeSymbol => TryCreateTypeForINamedTypeSymbol(namedTypeSymbol, validator, out type),
+
+                // We can only handle IArrayTypeSymbol of framework type and INamedTypeSymbol for now since CSharpType can't represent other types such as IArrayTypeSymbol of user types
+                // Instead of throwing an exception, wihch causes more side effects, we just return false and let the caller handle it.
+                _ => false
```

Comment:
maybe we could throw here as a emphasizer for us to update the code to handle more cases because returning false here for an unhandled case might be hiding issues.
But I could go either way here